cmd/ncproxy/service.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+const serviceName = "ncproxy"
+
+type handler struct {
+	fromsvc  chan error
+	s        *server
+	serveErr chan error
+}
+
+func getServicePath() (string, error) {
+	p, err := exec.LookPath(os.Args[0])
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(p)
+}
+
+type serviceFailureActions struct {
+	ResetPeriod  uint32
+	RebootMsg    *uint16
+	Command      *uint16
+	ActionsCount uint32
+	Actions      uintptr
+}
+
+type scAction struct {
+	Type  uint32
+	Delay uint32
+}
+
+// See http://stackoverflow.com/questions/35151052/how-do-i-configure-failure-actions-of-a-windows-service-written-in-go
+const (
+	scActionNone    = 0
+	scActionRestart = 1
+
+	serviceConfigFailureActions = 2
+)
+
+func registerService() error {
+	p, err := getServicePath()
+	if err != nil {
+		return err
+	}
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = m.Disconnect()
+	}()
+
+	c := mgr.Config{
+		ServiceType:  windows.SERVICE_WIN32_OWN_PROCESS,
+		StartType:    mgr.StartAutomatic,
+		ErrorControl: mgr.ErrorNormal,
+		DisplayName:  "Ncproxy",
+		Description:  "Network configuration proxy",
+	}
+
+	// Configure the service to launch with the arguments that were just passed.
+	args := []string{"-run-service"}
+	for _, a := range os.Args[1:] {
+		if a != "-register-service" && a != "-unregister-service" {
```

Comment:
isn't this only ever called when `-register-service` is set?

Reply:
Yea good point, the way those if statements are ordered right now this isn't needed.